### PR TITLE
Fix CCO related nits in `cirq.Operation` and `cirq.TaggedOperation`

### PR DESCRIPTION
**Pull request by tanujkhattar**
Asserts that:
  - `op.with_classical_controls() is op` : Returns self if set of conditions is empty. This is already true for other methods like `op.with_tags`, `op.controlled_by` etc.
  - `tagged_op.with_classical_controls()` will now remove tags, which is consistent with the philosophy that tagged operations are immutable and tags should be removed if the underlying operation is mutated. Other methods, like `tagged_op.controlled_by`, already follow this pattern.

### CODE REVIEW

**Review by 95-martin-orion — APPROVED** [review]
Overarching comment: prefer `if not some_list` instead of `if len(some_list) == 0`.

Otherwise this looks good to go.

File: cirq-core/cirq/ops/classically_controlled_operation_test.py [hunk 1]
```diff
@@ -385,7 +385,7 @@ def test_condition_removal():
     op = op.without_classical_controls()
     assert not cirq.control_keys(op)
     assert not op.classical_controls
-    assert set(map(str, op.tags)) == {'t1'}
+    assert not op.tags
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/classically_controlled_operation_test.py:388]: @95-martin-orion This changes the assumption that tags are preserved when calling `with_classical_controls`. This looks reasonable to me; since even in the original version, only `t1` was preserved instead of both `t1` and `t2`. 

Just want to make sure this looks good to you.

**Review by tanujkhattar — COMMENTED** [review]
@95-martin-orion Made a couple of changes to make the tests pass. PTAL and I'll merge once you re-approve.

File: cirq-core/cirq/ops/classically_controlled_operation_test.py (hunk 1 above)
Comment (daxfohl) [reply cirq-core/cirq/ops/classically_controlled_operation_test.py:388]: The docstrings for with/without_classical_controls both describe how tags are handled. If that behavior needs to change then the docstrings need updated to be consistent. (Probably can just change "hide" to "remove" in `with`, and I guess you can delete the blurb in `without` completely since it's no longer applicable).

File: cirq-core/cirq/ops/raw_types.py [hunk 2]
```diff
@@ -864,6 +868,13 @@ def without_classical_controls(self) -> 'cirq.Operation':
         new_sub_operation = self.sub_operation.without_classical_controls()
         return self if new_sub_operation is self.sub_operation else new_sub_operation
 
+    def with_classical_controls(
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply cirq-core/cirq/ops/raw_types.py:871]: I think that's fine, and also consistent with other operations like controlled ops. The general guidance would also be to use `op.with_classical_controls` instead of using the constructor.

[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/ops/raw_types.py:871]: sg, I guess the `without` docstring doesn't need updated then, unless there's a need to be explicit that that's the only case in which a TaggedOp could be returned. (`with` still needs updated though).

File: cirq-core/cirq/ops/classically_controlled_operation_test.py (hunk 1 above)
Comment (95-martin-orion) [reply cirq-core/cirq/ops/classically_controlled_operation_test.py:388]: +1 to Dax's comment, other than that I'm on board. Might want a BREAKING_CHANGE label to keep track of this, but I see it as a bugfix - the intent is for _any_ non-tagging operation to eliminate tags.

Comment (tanujkhattar) [reply cirq-core/cirq/ops/classically_controlled_operation_test.py:388]: Updated the docstring.

**Review by tanujkhattar — COMMENTED** [review]
Thanks for the feedback. 
Merging now to unblock the next PR, which will fix https://github.com/quantumlib/Cirq/issues/4977

File: cirq-core/cirq/ops/raw_types.py [hunk 3]
```diff
@@ -630,16 +630,19 @@ def with_classical_controls(
         measurement key. A measurement key evaluates to True iff any qubit in
         the corresponding measurement operation evaluated to a non-zero value.
 
-        The classical control will hide any tags on the existing operation,
-        since tags are considered a local attribute.
+        If no conditions are specified, returns self.
+
+        The classical control will remove any tags on the existing operation,
+        since tagged operations are considered to be immutable.
```
Comment (95-martin-orion) [inline cirq-core/cirq/ops/raw_types.py:636]: The reason isn't that TaggedOperations are immutable, as this creates a new operation - no mutation is taking place regardless. The actual reason is that tags are fragile - for an arbitrary tag, it's not possible to determine whether it should be preserved after a change, so we opt to always discard them.

Comment (tanujkhattar) [reply cirq-core/cirq/ops/raw_types.py:636]: Yeah, that's why I tried writing "considered" to be immutable. 

How about I reword it to?

```
The classical control will remove any tags on the existing operation,
since tags are fragile, and we always opt to get rid of the tags when
the underlying operation is changed.
```

Any suggestions with the right wording are welcome.

Comment (95-martin-orion) [reply cirq-core/cirq/ops/raw_types.py:636]: That looks good to me. Granting approval with the assumption that this gets applied.

**Review by 95-martin-orion — APPROVED** [review]
